rust-runtime/aws-smithy-runtime-api/src/client/result.rs

```diff
@@ -0,0 +1,657 @@
+    }
+
+    /// Returns the underlying service error `E` if there is one
+    ///
+    /// If the `SdkError` is not a `ServiceError` (for example, the error is a network timeout),
+    /// then it will be converted into an unhandled variant of `E`. This makes it easy to match
+    /// on the service's error response while simultaneously bubbling up transient failures.
+    /// For example, handling the `NoSuchKey` error for S3's `GetObject` operation may look as
+    /// follows:
+    ///
+    /// ```no_run
+    /// # use aws_smithy_runtime_api::client::result::{SdkError, CreateUnhandledError};
+    /// # #[derive(Debug)] enum GetObjectError { NoSuchKey(()), Other(()) }
+    /// # impl std::fmt::Display for GetObjectError {
+    /// #     fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result { unimplemented!() }
+    /// # }
+    /// # impl std::error::Error for GetObjectError {}
+    /// # impl CreateUnhandledError for GetObjectError {
+    /// #     fn create_unhandled_error(
+    /// #         _: Box<dyn std::error::Error + Send + Sync + 'static>,
+    /// #         _: Option<aws_smithy_types::error::ErrorMetadata>,
+    /// #     ) -> Self { unimplemented!() }
+    /// # }
+    /// # fn example() -> Result<(), GetObjectError> {
+    /// # let sdk_err = SdkError::service_error(GetObjectError::NoSuchKey(()), ());
+    /// match sdk_err.into_service_error() {
+    ///     GetObjectError::NoSuchKey(_) => {
+    ///         // handle NoSuchKey
+    ///     }
+    ///     err @ _ => return Err(err),
+    /// }
+    /// # Ok(())
+    /// # }
+    /// ```
+    pub fn into_service_error(self) -> E
+    where
+        E: std::error::Error + Send + Sync + CreateUnhandledError + 'static,
+        R: Debug + Send + Sync + 'static,
+    {
+        match self {
+            Self::ServiceError(context) => context.source,
+            _ => E::create_unhandled_error(self.into(), None),
+        }
+    }
+
+    /// Converts this error into its error source.
+    ///
+    /// If there is no error source, then `Err(Self)` is returned.
+    pub fn into_source(self) -> Result<Box<dyn Error + Send + Sync + 'static>, Self>
+    where
+        E: std::error::Error + Send + Sync + 'static,
+    {
+        use SdkError::*;
+        match self {
+            ConstructionFailure(context) => Ok(context.source),
+            TimeoutError(context) => Ok(context.source),
+            ResponseError(context) => Ok(context.source),
+            DispatchFailure(context) => Ok(context.source.into()),
+            ServiceError(context) => Ok(context.source.into()),
+        }
+    }
+
+    /// Return a reference to this error's raw response, if it contains one. Otherwise, return `None`.
+    pub fn raw_response(&self) -> Option<&R> {
+        match self {
+            Self::ServiceError(inner) => Some(inner.raw()),
+            Self::ResponseError(inner) => Some(inner.raw()),
+            _ => None,
+        }
+    }
```

Comment:
This is a random chore, feel free to ignore.
We're specifying variants in several of these methods differently. If you wouldn't mind, could you pick one style and update things?

Reply:
Thanks for pointing out! Yes, they look arbitrary. 8f5ebfe updated match arms in the file so they are consistently prefixed with `SdkError::`.